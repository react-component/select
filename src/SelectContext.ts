import * as React from 'react';
import type { RawValueType, RenderNode } from './BaseSelect';
import type { FlattenOptionData } from './interface';
import type { BaseOptionType, FieldNames, OnActiveValue, OnInternalSelect } from './Select';

// Use any here since we do not get the type during compilation
export interface SelectContextProps {
  options: BaseOptionType[];
  flattenOptions: FlattenOptionData<BaseOptionType>[];
  onActiveValue: OnActiveValue;
  defaultActiveFirstOption?: boolean;
  onSelect: OnInternalSelect;
  menuItemSelectedIcon?: RenderNode;
  rawValues: Set<RawValueType>;
  fieldNames?: FieldNames;
  virtual?: boolean;
  listHeight?: number;
  listItemHeight?: number;
  childrenAsData?: boolean;
}

const SelectContext = React.createContext<SelectContextProps>(null);

export default SelectContext;
