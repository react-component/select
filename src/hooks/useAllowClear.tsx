import TransBtn from '../TransBtn';
import type { DisplayValueType, Mode, RenderNode } from '../interface';
import React from 'react';

export const useAllowClear = (
  prefixCls: string,
  onClearMouseDown: React.MouseEventHandler<HTMLSpanElement>,
  displayValues: DisplayValueType[],
  allowClear?: boolean | { clearIcon?: RenderNode },
  clearIcon?: RenderNode,
  disabled: boolean = false,
  mergedSearchValue?: string,
  mode?: Mode,
) => {
  const mergedClearIcon = React.useMemo(() => {
    if (typeof allowClear === 'object') {
      return allowClear.clearIcon;
    }
    if (clearIcon) {
      return clearIcon;
    }
  }, [allowClear, clearIcon]);

  const mergedAllowClear = React.useMemo<boolean>(() => {
    if (
      !disabled &&
      !!allowClear &&
      (displayValues.length || mergedSearchValue) &&
      !(mode === 'combobox' && mergedSearchValue === '')
    ) {
      return true;
    }
    return false;
  }, [allowClear, disabled, displayValues.length, mergedSearchValue, mode]);

  return {
    allowClear: mergedAllowClear,
    clearIcon: (
      <TransBtn
        className={`${prefixCls}-clear`}
        onMouseDown={onClearMouseDown}
        customizeIcon={mergedClearIcon}
      >
        Ã—
      </TransBtn>
    ),
  };
};
