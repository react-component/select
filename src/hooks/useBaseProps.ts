/**
 * BaseSelect provide some parsed data into context.
 * You can use this hooks to get them.
 */

import * as React from 'react';
import type { BaseSelectProps } from '../BaseSelect';

export interface BaseSelectContextProps extends BaseSelectProps {
  triggerOpen: boolean;
  multiple: boolean;
  toggleOpen: (open?: boolean) => void;
}

export const BaseSelectContext = React.createContext<BaseSelectContextProps>(null);

export default function useBaseProps() {
  return React.useContext(BaseSelectContext);
}
