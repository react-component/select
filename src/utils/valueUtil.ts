import warning from 'rc-util/lib/warning';
import type {
  OptionsType as SelectOptionsType,
  OptionData,
  OptionGroupData,
  FlattenOptionData,
  FieldNames,
} from '../interface';
import type {
  LabelValueType,
  FilterFunc,
  RawValueType,
  GetLabeledValue,
  DefaultValueType,
} from '../interface/generator';

import { toArray } from './commonUtil';

function getKey(data: OptionData | OptionGroupData, index: number) {
  const { key } = data;
  let value: RawValueType;

  if ('value' in data) {
    ({ value } = data);
  }

  if (key !== null && key !== undefined) {
    return key;
  }
  if (value !== undefined) {
    return value;
  }
  return `rc-index-key-${index}`;
}

export function fillFieldNames(fieldNames?: FieldNames) {
  const { label, value, options } = fieldNames || {};

  return {
    label: label || 'label',
    value: value || 'value',
    options: options || 'options',
  };
}

/**
 * Flat options into flatten list.
 * We use `optionOnly` here is aim to avoid user use nested option group.
 * Here is simply set `key` to the index if not provided.
 */
export function flattenOptions(
  options: SelectOptionsType,
  { fieldNames }: { fieldNames?: FieldNames } = {},
): FlattenOptionData[] {
  const flattenList: FlattenOptionData[] = [];

  const {
    label: fieldLabel,
    value: fieldValue,
    options: fieldOptions,
  } = fillFieldNames(fieldNames);

  function dig(list: SelectOptionsType, isGroupOption: boolean) {
    list.forEach((data) => {
      const label = data[fieldLabel];

      if (isGroupOption || !(fieldOptions in data)) {
        // Option
        flattenList.push({
          key: getKey(data, flattenList.length),
          groupOption: isGroupOption,
          data,
          label,
          value: data[fieldValue],
        });
      } else {
        // Option Group
        flattenList.push({
          key: getKey(data, flattenList.length),
          group: true,
          data,
          label,
        });

        dig(data[fieldOptions], true);
      }
    });
  }

  dig(options, false);

  return flattenList;
}

/**
 * Inject `props` into `option` for legacy usage
 */
function injectPropsWithOption<T>(option: T): T {
  const newOption = { ...option };
  if (!('props' in newOption)) {
    Object.defineProperty(newOption, 'props', {
      get() {
        warning(
          false,
          'Return type is option instead of Option instance. Please read value directly instead of reading from `props`.',
        );
        return newOption;
      },
    });
  }

  return newOption;
}

export function findValueOption(
  values: RawValueType[],
  options: FlattenOptionData[],
  { prevValueOptions = [] }: { prevValueOptions?: OptionData[] } = {},
): OptionData[] {
  const optionMap: Map<RawValueType, OptionData> = new Map();

  options.forEach(({ data, group, value }) => {
    if (!group) {
      // Check if match
      optionMap.set(value, data as OptionData);
    }
  });

  return values.map((val) => {
    let option = optionMap.get(val);

    // Fallback to try to find prev options
    if (!option) {
      option = {
        // eslint-disable-next-line no-underscore-dangle
        ...prevValueOptions.find((opt) => opt._INTERNAL_OPTION_VALUE_ === val),
      };
    }

    return injectPropsWithOption(option);
  });
}

export const getLabeledValue: GetLabeledValue<FlattenOptionData[]> = (
  value,
  { options, prevValueMap, labelInValue, optionLabelProp },
): LabelValueType => {
  const item = findValueOption([value], options)[0];
  const result: LabelValueType = {
    value,
  };

  const prevValItem: LabelValueType = labelInValue ? prevValueMap.get(value) : undefined;

  if (prevValItem && typeof prevValItem === 'object' && 'label' in prevValItem) {
    result.label = prevValItem.label;

    if (
      item &&
      typeof prevValItem.label === 'string' &&
      typeof item[optionLabelProp] === 'string' &&
      prevValItem.label.trim() !== item[optionLabelProp].trim()
    ) {
      warning(false, '`label` of `value` is not same as `label` in Select options.');
    }
  } else if (item && optionLabelProp in item) {
    result.label = item[optionLabelProp];
  } else {
    result.label = value;
    result.isCacheable = true;
  }

  // Used for motion control
  result.key = result.value;

  return result;
};

function toRawString(content: React.ReactNode): string {
  return toArray(content).join('');
}

/** Filter single option if match the search text */
function getFilterFunction(optionFilterProp: string) {
  return (searchValue: string, option: OptionData | OptionGroupData) => {
    const lowerSearchText = searchValue.toLowerCase();

    // Group label search
    if ('options' in option) {
      return toRawString(option.label).toLowerCase().includes(lowerSearchText);
    }

    // Option value search
    const rawValue = option[optionFilterProp];
    const value = toRawString(rawValue).toLowerCase();
    return value.includes(lowerSearchText);
  };
}

/** Filter options and return a new options by the search text */
export function filterOptions(
  searchValue: string,
  options: SelectOptionsType,
  {
    optionFilterProp,
    filterOption,
  }: { optionFilterProp: string; filterOption: boolean | FilterFunc<SelectOptionsType[number]> },
) {
  const filteredOptions: SelectOptionsType = [];
  let filterFunc: FilterFunc<SelectOptionsType[number]>;

  if (filterOption === false) {
    return [...options];
  }
  if (typeof filterOption === 'function') {
    filterFunc = filterOption;
  } else {
    filterFunc = getFilterFunction(optionFilterProp);
  }

  options.forEach((item) => {
    // Group should check child options
    if ('options' in item) {
      // Check group first
      const matchGroup = filterFunc(searchValue, item);
      if (matchGroup) {
        filteredOptions.push(item);
      } else {
        // Check option
        const subOptions = item.options.filter((subItem) => filterFunc(searchValue, subItem));
        if (subOptions.length) {
          filteredOptions.push({
            ...item,
            options: subOptions,
          });
        }
      }

      return;
    }

    if (filterFunc(searchValue, injectPropsWithOption(item))) {
      filteredOptions.push(item);
    }
  });

  return filteredOptions;
}

export function getSeparatedContent(text: string, tokens: string[]): string[] {
  if (!tokens || !tokens.length) {
    return null;
  }

  let match = false;

  function separate(str: string, [token, ...restTokens]: string[]) {
    if (!token) {
      return [str];
    }

    const list = str.split(token);
    match = match || list.length > 1;

    return list
      .reduce((prevList, unitStr) => [...prevList, ...separate(unitStr, restTokens)], [])
      .filter((unit) => unit);
  }

  const list = separate(text, tokens);
  return match ? list : null;
}

export function isValueDisabled(value: RawValueType, options: FlattenOptionData[]): boolean {
  const option = findValueOption([value], options)[0];
  return option.disabled;
}

/**
 * `tags` mode should fill un-list item into the option list
 */
export function fillOptionsWithMissingValue(
  options: SelectOptionsType,
  value: DefaultValueType,
  optionLabelProp: string,
  labelInValue: boolean,
): SelectOptionsType {
  const values = toArray<RawValueType | LabelValueType>(value).slice().sort();
  const cloneOptions = [...options];

  // Convert options value to set
  const optionValues = new Set<RawValueType>();
  options.forEach((opt) => {
    if (opt.options) {
      opt.options.forEach((subOpt: OptionData) => {
        optionValues.add(subOpt.value);
      });
    } else {
      optionValues.add((opt as OptionData).value);
    }
  });

  // Fill missing value
  values.forEach((item) => {
    const val: RawValueType = labelInValue
      ? (item as LabelValueType).value
      : (item as RawValueType);

    if (!optionValues.has(val)) {
      cloneOptions.push(
        labelInValue
          ? {
              [optionLabelProp]: (item as LabelValueType).label,
              value: val,
            }
          : { value: val },
      );
    }
  });

  return cloneOptions;
}
