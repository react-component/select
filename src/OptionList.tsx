import * as React from 'react';
import { useEffect } from 'react';
import KeyCode from 'rc-util/lib/KeyCode';
import omit from 'rc-util/lib/omit';
import pickAttrs from 'rc-util/lib/pickAttrs';
import useMemo from 'rc-util/lib/hooks/useMemo';
import classNames from 'classnames';
import type { ListRef } from 'rc-virtual-list';
import List from 'rc-virtual-list';
import TransBtn from './TransBtn';
import type {
  OptionsType as SelectOptionsType,
  FlattenOptionData as SelectFlattenOptionData,
  OptionData,
  RenderNode,
  OnActiveValue,
  FieldNames,
} from './interface';
import type { RawValueType, FlattenOptionsType } from './interface/generator';
import { fillFieldNames } from './utils/valueUtil';
import { isPlatformMac } from './utils/platformUtil';

export interface OptionListProps<OptionsType extends object[]> {
  prefixCls: string;
  id: string;
  options: OptionsType;
  fieldNames?: FieldNames;
  flattenOptions: FlattenOptionsType<OptionsType>;
  height: number;
  itemHeight: number;
  values: Set<RawValueType>;
  multiple: boolean;
  open: boolean;
  defaultActiveFirstOption?: boolean;
  notFoundContent?: React.ReactNode;
  menuItemSelectedIcon?: RenderNode;
  childrenAsData: boolean;
  searchValue: string;
  virtual: boolean;
  direction?: 'ltr' | 'rtl';

  onSelect: (value: RawValueType, option: { selected: boolean }) => void;
  onToggleOpen: (open?: boolean) => void;
  /** Tell Select that some value is now active to make accessibility work */
  onActiveValue: OnActiveValue;
  onScroll: React.UIEventHandler<HTMLDivElement>;

  /** Tell Select that mouse enter the popup to force re-render */
  onMouseEnter?: React.MouseEventHandler;
}

export interface RefOptionListProps {
  onKeyDown: React.KeyboardEventHandler;
  onKeyUp: React.KeyboardEventHandler;
  scrollTo?: (index: number) => void;
}

/**
 * Using virtual list of option display.
 * Will fallback to dom if use customize render.
 */
const OptionList: React.ForwardRefRenderFunction<
  RefOptionListProps,
  OptionListProps<SelectOptionsType>
> = (
  {
    prefixCls,
    id,
    fieldNames,
    flattenOptions,
    childrenAsData,
    values,
    searchValue,
    multiple,
    defaultActiveFirstOption,
    height,
    itemHeight,
    notFoundContent,
    open,
    menuItemSelectedIcon,
    virtual,
    onSelect,
    onToggleOpen,
    onActiveValue,
    onScroll,
    onMouseEnter,
  },
  ref,
) => {
  const itemPrefixCls = `${prefixCls}-item`;

  const memoFlattenOptions = useMemo(
    () => flattenOptions,
    [open, flattenOptions],
    (prev, next) => next[0] && prev[1] !== next[1],
  );

  // =========================== List ===========================
  const listRef = React.useRef<ListRef>(null);

  const onListMouseDown: React.MouseEventHandler<HTMLDivElement> = (event) => {
    event.preventDefault();
  };

  const scrollIntoView = (index: number) => {
    if (listRef.current) {
      listRef.current.scrollTo({ index });
    }
  };

  // ========================== Active ==========================
  const getEnabledActiveIndex = (index: number, offset: number = 1): number => {
    const len = memoFlattenOptions.length;

    for (let i = 0; i < len; i += 1) {
      const current = (index + i * offset + len) % len;

      const { group, data } = memoFlattenOptions[current];
      if (!group && !(data as OptionData).disabled) {
        return current;
      }
    }

    return -1;
  };

  const [activeIndex, setActiveIndex] = React.useState(() => getEnabledActiveIndex(0));
  const setActive = (index: number, fromKeyboard = false) => {
    setActiveIndex(index);

    const info = { source: fromKeyboard ? ('keyboard' as const) : ('mouse' as const) };

    // Trigger active event
    const flattenItem = memoFlattenOptions[index];
    if (!flattenItem) {
      onActiveValue(null, -1, info);
      return;
    }

    onActiveValue((flattenItem.data as OptionData).value, index, info);
  };

  // Auto active first item when list length or searchValue changed
  useEffect(() => {
    setActive(defaultActiveFirstOption !== false ? getEnabledActiveIndex(0) : -1);
  }, [memoFlattenOptions.length, searchValue]);

  // Auto scroll to item position in single mode
  useEffect(() => {
    /**
     * React will skip `onChange` when component update.
     * `setActive` function will call root accessibility state update which makes re-render.
     * So we need to delay to let Input component trigger onChange first.
     */
    const timeoutId = setTimeout(() => {
      if (!multiple && open && values.size === 1) {
        const value: RawValueType = Array.from(values)[0];
        const index = memoFlattenOptions.findIndex(
          ({ data }) => (data as OptionData).value === value,
        );

        if (index !== -1) {
          setActive(index);
          scrollIntoView(index);
        }
      }
    });

    // Force trigger scrollbar visible when open
    if (open) {
      listRef.current?.scrollTo(undefined);
    }

    return () => clearTimeout(timeoutId);
  }, [open, searchValue]);

  // ========================== Values ==========================
  const onSelectValue = (value: RawValueType) => {
    if (value !== undefined) {
      onSelect(value, { selected: !values.has(value) });
    }

    // Single mode should always close by select
    if (!multiple) {
      onToggleOpen(false);
    }
  };

  // ========================= Keyboard =========================
  React.useImperativeHandle(ref, () => ({
    onKeyDown: (event) => {
      const { which, ctrlKey } = event;
      switch (which) {
        // >>> Arrow keys & ctrl + n/p on Mac
        case KeyCode.N:
        case KeyCode.P:
        case KeyCode.UP:
        case KeyCode.DOWN: {
          let offset = 0;
          if (which === KeyCode.UP) {
            offset = -1;
          } else if (which === KeyCode.DOWN) {
            offset = 1;
          } else if (isPlatformMac() && ctrlKey) {
            if (which === KeyCode.N) {
              offset = 1;
            } else if (which === KeyCode.P) {
              offset = -1;
            }
          }

          if (offset !== 0) {
            const nextActiveIndex = getEnabledActiveIndex(activeIndex + offset, offset);
            scrollIntoView(nextActiveIndex);
            setActive(nextActiveIndex, true);
          }

          break;
        }

        // >>> Select
        case KeyCode.ENTER: {
          // value
          const item = memoFlattenOptions[activeIndex];
          if (item && !(item.data as OptionData).disabled) {
            onSelectValue((item.data as OptionData).value);
          } else {
            onSelectValue(undefined);
          }

          if (open) {
            event.preventDefault();
          }

          break;
        }

        // >>> Close
        case KeyCode.ESC: {
          onToggleOpen(false);
          if (open) {
            event.stopPropagation();
          }
        }
      }
    },
    onKeyUp: () => {},

    scrollTo: (index) => {
      scrollIntoView(index);
    },
  }));

  // ========================== Render ==========================
  if (memoFlattenOptions.length === 0) {
    return (
      <div
        role="listbox"
        id={`${id}_list`}
        className={`${itemPrefixCls}-empty`}
        onMouseDown={onListMouseDown}
      >
        {notFoundContent}
      </div>
    );
  }

  const filledFieldNames = fillFieldNames(fieldNames);
  const omitFieldNameList = Object.keys(filledFieldNames).map((key) => filledFieldNames[key]);

  const renderItem = (index: number) => {
    const item = memoFlattenOptions[index];
    if (!item) return null;

    const itemData = (item.data || {}) as OptionData;
    const { value, label, children } = itemData;
    const attrs = pickAttrs(itemData, true);
    const mergedLabel = childrenAsData ? children : label;
    return item ? (
      <div
        aria-label={typeof mergedLabel === 'string' ? mergedLabel : null}
        {...attrs}
        key={index}
        role="option"
        id={`${id}_list_${index}`}
        aria-selected={values.has(value)}
      >
        {value}
      </div>
    ) : null;
  };

  return (
    <>
      <div role="listbox" id={`${id}_list`} style={{ height: 0, width: 0, overflow: 'hidden' }}>
        {renderItem(activeIndex - 1)}
        {renderItem(activeIndex)}
        {renderItem(activeIndex + 1)}
      </div>
      <List<SelectFlattenOptionData>
        itemKey="key"
        ref={listRef}
        data={memoFlattenOptions}
        height={height}
        itemHeight={itemHeight}
        fullHeight={false}
        onMouseDown={onListMouseDown}
        onScroll={onScroll}
        virtual={virtual}
        onMouseEnter={onMouseEnter}
      >
        {({ group, groupOption, data, label, value }, itemIndex) => {
          const { key } = data;

          // Group
          if (group) {
            return (
              <div className={classNames(itemPrefixCls, `${itemPrefixCls}-group`)}>
                {label !== undefined ? label : key}
              </div>
            );
          }

          const { disabled, title, children, style, className, ...otherProps } = data as OptionData;
          const passedProps = omit(otherProps, omitFieldNameList);

          // Option
          const selected = values.has(value);

          const optionPrefixCls = `${itemPrefixCls}-option`;
          const optionClassName = classNames(itemPrefixCls, optionPrefixCls, className, {
            [`${optionPrefixCls}-grouped`]: groupOption,
            [`${optionPrefixCls}-active`]: activeIndex === itemIndex && !disabled,
            [`${optionPrefixCls}-disabled`]: disabled,
            [`${optionPrefixCls}-selected`]: selected,
          });

          const mergedLabel = childrenAsData ? children : label;

          const iconVisible =
            !menuItemSelectedIcon || typeof menuItemSelectedIcon === 'function' || selected;

          const content = mergedLabel || value;
          // https://github.com/ant-design/ant-design/issues/26717
          let optionTitle =
            typeof content === 'string' || typeof content === 'number'
              ? content.toString()
              : undefined;
          if (title !== undefined) {
            optionTitle = title;
          }

          return (
            <div
              {...passedProps}
              aria-selected={selected}
              className={optionClassName}
              title={optionTitle}
              onMouseMove={() => {
                if (activeIndex === itemIndex || disabled) {
                  return;
                }
                setActive(itemIndex);
              }}
              onClick={() => {
                if (!disabled) {
                  onSelectValue(value);
                }
              }}
              style={style}
            >
              <div className={`${optionPrefixCls}-content`}>{content}</div>
              {React.isValidElement(menuItemSelectedIcon) || selected}
              {iconVisible && (
                <TransBtn
                  className={`${itemPrefixCls}-option-state`}
                  customizeIcon={menuItemSelectedIcon}
                  customizeIconProps={{ isSelected: selected }}
                >
                  {selected ? '✓' : null}
                </TransBtn>
              )}
            </div>
          );
        }}
      </List>
    </>
  );
};

const RefOptionList = React.forwardRef<RefOptionListProps, OptionListProps<SelectOptionsType>>(
  OptionList,
);
RefOptionList.displayName = 'OptionList';

export default RefOptionList;
