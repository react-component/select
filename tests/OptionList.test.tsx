import { mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { BaseSelectContext } from '../src/hooks/useBaseProps';
import type { RefOptionListProps } from '../src/OptionList';
import OptionList from '../src/OptionList';
import SelectContext from '../src/SelectContext';
import { fillFieldNames, flattenOptions } from '../src/utils/valueUtil';
import { injectRunAllTimers } from './utils/common';

jest.mock('../src/utils/platformUtil');

describe('OptionList', () => {
  injectRunAllTimers(jest);

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  function generateList({ options, values, ref, ...props }: any) {
    const fieldNames = fillFieldNames(props.fieldNames || {}, false);
    const flattenedOptions = flattenOptions(options, {
      fieldNames,
      childrenAsData: false,
    });

    return (
      <BaseSelectContext.Provider
        value={{
          prefixCls: 'rc-select',
          ...props,
        }}
      >
        <SelectContext.Provider
          value={{
            fieldNames,
            flattenOptions: flattenedOptions,
            options,
            onActiveValue: () => {},
            onSelect: () => {},
            rawValues: values || new Set(),
            virtual: true,
            ...props,
          }}
        >
          <div>
            <OptionList ref={ref} />
          </div>
        </SelectContext.Provider>
      </BaseSelectContext.Provider>
    );
  }

  describe('renders correctly', () => {
    const sharedConfig = {
      options: [
        {
          key: 'group1',
          options: [{ value: '1', 'aria-label': 'value-1' }],
        },
        {
          key: 'group2',
          options: [{ value: '2' }],
        },
      ],
      values: new Set(['1']),
    };

    it('virtual', () => {
      const wrapper = mount(generateList(sharedConfig));
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('without virtual', () => {
      const wrapper = mount(
        generateList({
          ...sharedConfig,
          virtual: false,
        }),
      );
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  it('save first active item', () => {
    const onActiveValue = jest.fn();

    mount(
      generateList({
        options: [{ value: '1' }, { value: '2' }],
        values: new Set('1'),
        onActiveValue,
      }),
    );

    expect(onActiveValue).toHaveBeenCalledWith('1', expect.anything(), expect.anything());
  });

  it('key operation', () => {
    const onActiveValue = jest.fn();
    const listRef = React.createRef<RefOptionListProps>();
    mount(
      generateList({
        options: [{ value: '1' }, { value: '2' }],
        onActiveValue,
        ref: listRef,
      }),
    );

    onActiveValue.mockReset();
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.DOWN } as any);
    });
    expect(onActiveValue).toHaveBeenCalledWith(
      '2',
      expect.anything(),
      expect.objectContaining({ source: 'keyboard' }),
    );

    onActiveValue.mockReset();
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.UP } as any);
    });
    expect(onActiveValue).toHaveBeenCalledWith(
      '1',
      expect.anything(),
      expect.objectContaining({ source: 'keyboard' }),
    );
  });

  it('key operation with fieldNames', () => {
    const onActiveValue = jest.fn();
    const toggleOpen = jest.fn();
    const onSelect = jest.fn();
    const listRef = React.createRef<RefOptionListProps>();
    mount(
      generateList({
        fieldNames: { value: 'foo', label: 'bar' },
        options: [{ foo: '1' }, { foo: '2' }],
        onActiveValue,
        onSelect,
        toggleOpen,
        ref: listRef,
      }),
    );

    onActiveValue.mockReset();
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.DOWN } as any);
    });
    expect(onActiveValue).toHaveBeenCalledWith(
      '2',
      expect.anything(),
      expect.objectContaining({ source: 'keyboard' }),
    );

    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.ENTER } as any);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith('2', expect.objectContaining({ selected: true }));

    onSelect.mockReset();
    onActiveValue.mockReset();
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.UP } as any);
    });
    expect(onActiveValue).toHaveBeenCalledWith(
      '1',
      expect.anything(),
      expect.objectContaining({ source: 'keyboard' }),
    );

    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.ENTER } as any);
    });
    expect(onSelect).toHaveBeenCalledTimes(1);
    expect(onSelect).toHaveBeenCalledWith('1', expect.objectContaining({ selected: true }));
  });

  // mocked how we detect running platform in test environment
  it('special key operation on Mac', () => {
    const onActiveValue = jest.fn();
    const listRef = React.createRef<RefOptionListProps>();
    mount(
      generateList({
        options: [{ value: '1' }, { value: '2' }],
        onActiveValue,
        ref: listRef,
      }),
    );

    onActiveValue.mockReset();
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.N, ctrlKey: true } as any);
    });
    expect(onActiveValue).toHaveBeenCalledWith(
      '2',
      expect.anything(),
      expect.objectContaining({ source: 'keyboard' }),
    );

    onActiveValue.mockReset();
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.P, ctrlKey: true } as any);
    });
    expect(onActiveValue).toHaveBeenCalledWith(
      '1',
      expect.anything(),
      expect.objectContaining({ source: 'keyboard' }),
    );
  });

  it('hover to active', () => {
    const onActiveValue = jest.fn();
    const wrapper = mount(
      generateList({
        options: [{ value: '1' }, { value: '2' }],
        onActiveValue,
      }),
    );

    onActiveValue.mockReset();
    wrapper.find('.rc-select-item-option').last().simulate('mouseMove');
    expect(onActiveValue).toHaveBeenCalledWith(
      '2',
      expect.anything(),
      expect.objectContaining({ source: 'mouse' }),
    );

    // Same item not repeat trigger
    onActiveValue.mockReset();
    wrapper.find('.rc-select-item-option').last().simulate('mouseMove');
    expect(onActiveValue).not.toHaveBeenCalled();
  });

  it('Should prevent default with list mouseDown to avoid losing focus', () => {
    const wrapper = mount(
      generateList({
        options: [{ value: '1' }, { value: '2' }],
      }),
    );

    const preventDefault = jest.fn();
    wrapper.find('.rc-select-item-option').last().simulate('mouseDown', {
      preventDefault,
    });

    expect(preventDefault).toHaveBeenCalled();
  });

  it('Data attributes should be set correct', () => {
    const wrapper = mount(
      generateList({
        options: [
          { value: '1', label: 'my-label' },
          { value: '2', 'data-num': '123' },
        ],
      }),
    );

    expect(wrapper.find('.rc-select-item-option').last().prop('data-num')).toBe('123');
  });

  it('should render title defaultly', () => {
    const wrapper = mount(
      generateList({
        options: [{ value: '1', label: 'my-label' }],
      }),
    );
    expect(wrapper.find('.rc-select-item-option').first().prop('title')).toBe('my-label');
  });

  it('should render title', () => {
    const wrapper = mount(
      generateList({
        options: [{ value: '1', label: 'my-label', title: 'title' }],
      }),
    );
    expect(wrapper.find('.rc-select-item-option').first().prop('title')).toBe('title');
  });

  it('should not render title when title is empty string', () => {
    const wrapper = mount(
      generateList({
        options: [{ value: '1', label: 'my-label', title: '' }],
      }),
    );
    expect(wrapper.find('.rc-select-item-option').first().prop('title')).toBe('');
  });

  it('should render title from label when title is undefined', () => {
    const wrapper = mount(
      generateList({
        options: [{ value: '1', label: 'my-label', title: undefined }],
      }),
    );
    expect(wrapper.find('.rc-select-item-option').first().prop('title')).toBe('my-label');
  });

  it('should not render title defaultly when label is ReactNode', () => {
    const wrapper = mount(
      generateList({
        options: [{ value: '1', label: <div>label</div> }],
      }),
    );
    expect(wrapper.find('.rc-select-item-option').first().prop('title')).toBe(undefined);
  });

  it('params to scrollIntoView should be object when key is pressed', () => {
    const listRef = React.createRef<RefOptionListProps>();
    const options = Array.from({ length: 20 }).map((v, i) => ({ value: i, label: i }));
    const mockScroll = jest.fn();
    const wrapper = mount(
      generateList({
        options,
        ref: listRef,
      }),
    );
    const vList = wrapper.find('List').getElement() as any;
    // override the ref to test the parameters passed to the scrollIntoView method
    vList.ref.current = { scrollTo: mockScroll };
    act(() => {
      listRef.current.onKeyDown({ which: KeyCode.DOWN } as any);
    });
    expect(mockScroll.mock.calls[0][0]).toEqual({ index: 1 });
  });
});
