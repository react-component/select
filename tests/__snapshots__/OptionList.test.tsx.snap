// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionList renders correctly virtual 1`] = `
<div>
  <div
    id="undefined_list"
    role="listbox"
    style="height: 0px; width: 0px; overflow: hidden;"
  >
    <div
      aria-selected="false"
      id="undefined_list_0"
      role="presentation"
    />
    <div
      aria-label="value-1"
      aria-selected="true"
      id="undefined_list_1"
      role="option"
    >
      1
    </div>
    <div
      aria-selected="false"
      id="undefined_list_2"
      role="presentation"
    />
  </div>
  <div
    class="rc-virtual-list"
    style="position: relative;"
  >
    <div
      class="rc-virtual-list-holder"
    >
      <div>
        <div
          class="rc-virtual-list-holder-inner"
          style="display: flex; flex-direction: column;"
        >
          <div
            class="rc-select-item rc-select-item-group"
          >
            group1
          </div>
          <div
            aria-label="value-1"
            aria-selected="true"
            class="rc-select-item rc-select-item-option rc-select-item-option-grouped rc-select-item-option-active rc-select-item-option-selected"
            title="1"
          >
            <div
              class="rc-select-item-option-content"
            >
              1
            </div>
            <span
              aria-hidden="true"
              class="rc-select-item-option-state"
              style="user-select: none;"
              unselectable="on"
            >
              <span
                class="rc-select-item-option-state-icon"
              >
                ✓
              </span>
            </span>
          </div>
          <div
            class="rc-select-item rc-select-item-group"
          >
            group2
          </div>
          <div
            aria-selected="false"
            class="rc-select-item rc-select-item-option rc-select-item-option-grouped"
            title="2"
          >
            <div
              class="rc-select-item-option-content"
            >
              2
            </div>
            <span
              aria-hidden="true"
              class="rc-select-item-option-state"
              style="user-select: none;"
              unselectable="on"
            >
              <span
                class="rc-select-item-option-state-icon"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OptionList renders correctly without virtual 1`] = `
<div>
  <div
    class="rc-virtual-list"
    style="position: relative;"
  >
    <div
      class="rc-virtual-list-holder"
    >
      <div>
        <div
          class="rc-virtual-list-holder-inner"
          id="undefined_list"
          role="listbox"
          style="display: flex; flex-direction: column;"
        >
          <div
            class="rc-select-item rc-select-item-group"
          >
            group1
          </div>
          <div
            aria-label="value-1"
            aria-selected="true"
            class="rc-select-item rc-select-item-option rc-select-item-option-grouped rc-select-item-option-active rc-select-item-option-selected"
            id="undefined_list_1"
            role="option"
            title="1"
          >
            <div
              class="rc-select-item-option-content"
            >
              1
            </div>
            <span
              aria-hidden="true"
              class="rc-select-item-option-state"
              style="user-select: none;"
              unselectable="on"
            >
              <span
                class="rc-select-item-option-state-icon"
              >
                ✓
              </span>
            </span>
          </div>
          <div
            class="rc-select-item rc-select-item-group"
          >
            group2
          </div>
          <div
            aria-selected="false"
            class="rc-select-item rc-select-item-option rc-select-item-option-grouped"
            id="undefined_list_3"
            role="option"
            title="2"
          >
            <div
              class="rc-select-item-option-content"
            >
              2
            </div>
            <span
              aria-hidden="true"
              class="rc-select-item-option-state"
              style="user-select: none;"
              unselectable="on"
            >
              <span
                class="rc-select-item-option-state-icon"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
